src/ucp/core/ucp_worker.c

```diff
@@ -1499,28 +1498,27 @@ ucs_status_t ucp_worker_create(ucp_context_h context,
 
     if (params->field_mask & UCP_WORKER_PARAM_FIELD_THREAD_MODE) {
 #if !ENABLE_MT
-        thread_mode = UCS_THREAD_MODE_SINGLE;
+        uct_thread_mode = UCS_THREAD_MODE_SINGLE;
         if (params->thread_mode != UCS_THREAD_MODE_SINGLE) {
             ucs_debug("forced single thread mode on worker create");
         }
 #else
-        thread_mode = params->thread_mode;
-#endif
-    } else {
-        thread_mode = UCS_THREAD_MODE_SINGLE;
-    }
+        if (params->thread_mode != UCS_THREAD_MODE_SINGLE) {
+            /* UCT is serialized by UCP lock or by UCP user */
+            uct_thread_mode = UCS_THREAD_MODE_SERIALIZED;
+        } else {
+            uct_thread_mode = UCS_THREAD_MODE_SINGLE;
+        }
 
-    if (thread_mode == UCS_THREAD_MODE_MULTI) {
-        worker->flags = UCP_WORKER_FLAG_MT;
+        if (params->thread_mode == UCS_THREAD_MODE_MULTI) {
+            worker->flags = UCP_WORKER_FLAG_MT;
+        } else {
+            worker->flags = 0;
+        }
+#endif
     } else {
-        worker->flags = 0;
-    }
-
-    if (thread_mode == UCS_THREAD_MODE_SINGLE) {
+        worker->flags   = 0;
```

Comment:
maybe just initialize worker flags to 0 before all checks? and also remove line 1516

Reply:
good point - then comment above doesn't make sense
btw, it is already 0 as we call `ucs_calloc` for worker object

thank you!